segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -171,8 +171,18 @@ public void setupAppend(SetupAppend setupAppend) {
                                 if (u != null) {
                                     handleException(writer, setupAppend.getRequestId(), newSegment, "setting up append", u);
                                 } else {
+                                    // Last event number stored according to Segment store.
                                     long eventNumber = attributes.getOrDefault(writer, Attributes.NULL_ATTRIBUTE_VALUE);
-                                    this.writerStates.putIfAbsent(Pair.of(newSegment, writer), new WriterState(eventNumber));
+
+                                    // Create a new WriterState object based on the attribute value for the last event number for the writer.
+                                    // It should be noted that only one connection for a given segment writer is created by the client.
+                                    // The event number sent by the AppendSetup command is an implicit ack, the writer acks all events
+                                    // below the specified event number.
+                                    WriterState current = this.writerStates.put(Pair.of(newSegment, writer), new WriterState(eventNumber));
```

Comment:
We need to worry about a race here. There could be concurrent modification of the writer state. So we want to have an `update` function to check that we aren't replacing a newer value.

Reply:
Observed that all the append requests which are in pipeline, with requests to `store.Append(..)` sent before disconnect,  would have already fetched the reference to `WriterState` object. Hence adding a new `WriterState` object in response to a SetupAppend request by the writer on reconnect would imply that any updates to the previous writerState object would not affect the newly added writer state object.

Do let me know if I have missed any aspect here...

Reply:
Sounds good.